### ISSUE TITLE
Exception after first login with device streams proxy sample

### ISSUE THREAD

**Issue by dsajanice** [Azure-Samples/azure-iot-samples-csharp#38]
> Please provide us with the following information:
> ---------------------------------------------------------------

### This issue is for a: (mark with an `x`)
```
- [x] bug report -> please search issues before submitting
- [ ] feature request
- [ ] documentation issue or request
- [ ] regression (a behavior that used to work and stopped in a new release)
```

### Minimal steps to reproduce
>
1. Use device streams proxy samples to set up device and service side SSH proxies on the SAME machine (experiencing different issues cross machine).

2. Log in using SSH via service side proxy. See it work.

3. Exit SSH session.

4. Log in again using SSH via service side proxy. This second login throws an exception on the device proxy:

Microsoft.Azure.Devices.Common.Exceptions.DeviceNotFoundException: Device {"Message":"{\"errorCode\":404001,\"trackingId\":\"c45f6aee9d2b4cc78402ad05811ca501-G:11-TimeStamp:02/06/2019 22:41:10\",\"message\":\"Device unavailable\",\"timestampUtc\":\"2019-02-06T22:41:10.9080359Z\"}","ExceptionMessage":""} not registered
   at Microsoft.Azure.Devices.HttpClientHelper.ExecuteAsync(HttpClient httpClient, HttpMethod httpMethod, Uri requestUri, Func`3 modifyRequestMessageAsync, Func`2 isMappedToException, Func`3 processResponseMessageAsync, IDictionary`2 errorMappingOverrides, CancellationToken cancellationToken)
   at Microsoft.Azure.Devices.HttpClientHelper.PostAsync[T,T2](Uri requestUri, T entity, TimeSpan operationTimeout, IDictionary`2 errorMappingOverrides, IDictionary`2 customHeaders, CancellationToken cancellationToken)
   at Microsoft.Azure.Devices.AmqpServiceClient.CreateStreamAsync(Uri uri, DeviceStreamRequest deviceStreamRequest, CancellationToken cancellationToken)
   at Microsoft.Azure.Devices.Samples.DeviceStreamSample.HandleIncomingConnectionsAndCreateStreams(String deviceId, ServiceClient serviceClient, TcpClient tcpClient) in /home/jadsa-auto/Repos/azure-iot-samples-csharp/iot-hub/Quickstarts/device-streams-proxy/service/DeviceStreamSample.cs:line 60
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location where exception was thrown ---
   at System.Threading.ThreadPoolWorkQueue.Dispatch()

5. Observe that service side proxy terminates with no exception as well.

### Any log messages given by the failure
>

### Expected/desired behavior
> Multiple logins should work

### OS and Version?
> Ubuntu 18.04

### Versions
>

### Mention any other details that might be useful

> ---------------------------------------------------------------
> Thanks! We'll be in touch soon.

**Comment by dsajanice** [comment]
I modified the code to wait for another device stream after the first one was closed but I now hit the following exception after around 5 minutes:

Unhandled Exception: System.NullReferenceException: Object reference not set to an instance of an object.
   at Microsoft.Azure.Devices.Client.Transport.AmqpTransportHandler.WaitForDeviceStreamRequestAsync(CancellationToken cancellationToken)
   at Microsoft.Azure.Devices.Client.Transport.ErrorDelegatingHandler.ExecuteWithErrorHandlingAsync[T](Func`1 asyncOperation)
   at Microsoft.Azure.Devices.Client.Transport.RetryDelegatingHandler.<>c__DisplayClass25_0.<<WaitForDeviceStreamRequestAsync>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.Azure.Devices.Client.Transport.RetryDelegatingHandler.WaitForDeviceStreamRequestAsync(CancellationToken cancellationToken)
   at Microsoft.Azure.Devices.Client.InternalClient.WaitForDeviceStreamRequestAsync(CancellationToken cancellationToken)
   at Microsoft.Azure.Devices.Client.Samples.DeviceStreamSample.RunSampleAsync(Boolean acceptDeviceStreamingRequest) in /home/jadsa-auto/Repos/azure-iot-samples-csharp/iot-hub/Quickstarts/device-streams-proxy/device/DeviceStreamSample.cs:line 66
   at Microsoft.Azure.Devices.Client.Samples.DeviceStreamSample.RunSampleAsync() in /home/jadsa-auto/Repos/azure-iot-samples-csharp/iot-hub/Quickstarts/device-streams-proxy/device/DeviceStreamSample.cs:line 54
   at Microsoft.Azure.Devices.Client.Samples.Program.Main(String[] args) in /home/jadsa-auto/Repos/azure-iot-samples-csharp/iot-hub/Quickstarts/device-streams-proxy/device/Program.cs:line 81

Is there a baked in timeout in the SDK?

**Comment by rezasherafat** [comment]
Stack trace appears similar to issue #30.
@ewertons can you have a look?

**Comment by jenoola** [comment]
@dsajanice please specify the following:
1) OS you are using
2) how exactly you modified the sample
3) all steps to get repro
I do not see the issue on Win 10 with the following modification in Program.cs of service sample:
```
    using (ServiceClient serviceClient = ServiceClient.CreateFromConnectionString(s_connectionString, s_transportType))
            {
                var sample = new DeviceStreamSample(serviceClient, s_deviceId, port);
                sample.RunSampleAsync().GetAwaiter().GetResult();
                sample.RunSampleAsync().GetAwaiter().GetResult();
            }
```
The steps:
1. Start device proxy
2. Start service proxy
3. Connect to service proxy and close the connection - after this sample finishes, device proxy is closed and service proxy finishes the first RunSampleAsync
4. Start device proxy again
5. Connect to service proxy and close the connection - the same successful run as before, at the end both device and service proxy are closed with no error

**Comment by jenoola** [comment]
@dsajanice @rezasherafat do you still see the issue? If yes, please, provide details for your environment.

**Comment by aguywithcode** [comment]
I ran into the same issue. It appears that the Device Stream stays open after the the SSH client exits for some time. I believe the SSH service assumes that the client might come back but after some time finally closes the connection. I waited about five minutes after exiting the client and the local proxy finally gave a "done streaming" message and I was able to connect again.

I made some modifications to the code so that when a client attempts to connect, it first checks if there is a cancellation token source and cancels it before opening a new stream. I'm submitting a pr for this issue.
